Add missing features to SQLite runner

mattbrailsford opened the pull request:
This PR is looking at implementing most of the supported features of SQLite that appear to be missing in the current implementation.

Progress:
* [x] Added Schema support
* [x] Added SQLite drop column support (Fixes #812)
* [x] Add UNIQUE constraint support (Fixes #1492 and #662)
* [x] Switched to CompatabilityMode.Strict by default so that any missing features will warn users and not just fail silently (Fixes #1485)

Notes:
* UNIQUE constraint support is added in two ways. Either when defining the table you can use `.Unique()` which does a true UNIQUE constraint on the column or when using the `Create.UniqueConstraint` we actually just create a UNIQUE index instead.
* ~~Couldn't upgrade Microsoft.Data.Sqlite to v6 due to error `Unable to load DLL 'e_sqlite3'` (some background here https://github.com/dotnet/efcore/issues/19396) so upgraded to latest v5 where the error isn't present~~ Have now managed to upgrade to v6 by installing `SQLitePCLRaw.bundle_e_sqlite3` too.

mattbrailsford commented:
@jzabroski couple of things so far I could do with your feedback on.

1. In the migration test project, the `Microsoft.Data.Sqlite` dependency was for v3.1.5 which doesn't actually exist (3.1.24 is the highest v3). I'm currently needing to upgrade this to at least 5.0.16 in order to have support for some features implemented. Is there an objection to upgrading? And if so, should I aim for the lowest version possible? Or should I just go for the highest latest version which would be 6.0.4 (still ..NET Standard 2.0 so should be as compatible).
2. I've added in "Schema" support as technically SQLite supports it, but it's not exactly "schemas" as you would have in other providers. SQLite lets you attach other DB files into the current connection and alias them. You can then use the schema prefix to state which DB you are refering to. In adding in the schema support, I've broken some unit tests as I'm not sure how to add in an attached DB to support this. Any thoughts here? Should I not bother with schema support because we don't have an API for attaching DBs? Or should there be some kinda of way of attaching a DB and thus allow schema prefixes?

I've got more stuff to implement, but these are the initial things I could do with feedback on.

jzabroski commented:
Upgrading Microsoft.Data.Sqlite is fine. All your changes will go in 4.0 so SemVer practice is followed. 

I need to read docs on attaching multiple DB. I believe I wrote an ADR on Schema support. Check top level ADR folder. About to drive now.

mattbrailsford commented:
@jzabroski please see nots in the initial comment (I've been updating it) and see if you are happy with the comments.

If you are, I think it's really just what to do about the schema stuff which I think is one of 3 things

1. Just don't run those failing unit tests for Sqlite (not ideal)
2. Find a way to support attaching DB's to a current connection and then update the test for Sqlite to attach a DB with the Schema alias and then ensure the tests actually run (probably the preferred option)
3. We rollback the schema support and just say FM doesn't support schemas and if you need to run migrations across multiple DB's then you need to define multiple migrations + migration runners for each one.

Let me know what you'd prefer and also whether you have any issues with any of the changes.

mattbrailsford commented:
PS I've search the issue tracker for `SQLite` and can't find any other issues other than those addressed here 👍

jzabroski commented:
> 2. Find a way to support attaching DB's to a current connection and then update the test for Sqlite to attach a DB with the Schema alias and then ensure the tests actually run (probably the "correct" option)

What would this look like ? Did we prove that Microsoft.Data.Sqlite can do this outside FM? I think what you're saying is schema support is dependent on the current connection, almost like creating a linked server in other db products. 

I'm open to making connection management more robust. We have requests today to support Azure login tokens, for example

mattbrailsford commented:
I have got it working by calling

`Execute.Sql(“ATTACH DATABASE ‘’ TestSchema”) `

I can then run Schema based queries against this. This does allow me to move forward in some of the tests, but the move to CompatabilityMode.Strict is causing the remaining errors as I think all the Unit tests are relying on failing silently for missing features (Given Loose is the default for all other runners)

Not sure what’s best here.

mattbrailsford commented:
I'm wondering if we shouldn't revert back to loose compatibility, but in this mode log out the compatibility message so there is evidence that something is wrong (which is ultimately what the issue is really about that started this). It's still possible for people to run in strict mode if they wish by passing options to the migration runner.

mattbrailsford commented:
**UPDATE** I'm pretty close to having unit tests passing so will update once I have 👍

jzabroski commented:
> **UPDATE** In looking though, I'm not sure the best way to achieve logging from the generator 🤷‍♂️

Do you need help with this?

mattbrailsford commented:
@jzabroski ok, I've been through all the integration tests and fixed broken ones, but also enabled a whole bunch that were previously disabled for SQLite. I think there are one or two more (specifically rollback ones) that should be re-enabled, but because we are running tests on SQLite "in memory" I believe the uses of scopes / exceptions causes the in memory DB to reset and so we can't assert those tests.

I had a go at converting SQLite to use a db file stored in "Data Directory" but I couldn't get this to work (just kept getting a connection error) and so I just reverted this back. It does mean there are a couple of tests that could be run that aren't, but there are at least a heck of a lot more than were enabled before.

One interesting thing of done a bit of an awkward work around on is SQLite only supports FK's in the create table statement and whilst there was code to handle this in to create those FK definitions in the generated SQL, there didn't appear to be a way to prevent those FK's also being passed to the FK constraint generator and so I have to do some weird stuff, changing the name of the FK if it's one we've processed in the create table routing and if it is, don't throw the compatibility exception.

It works, but this could probably do with a specific API to know if a FK constraint expression is for a column in the create table setup, or if it's just an expressly defined FK expression.

Anyways, with this I think I can call this PR done and ready for you reviews.

mattbrailsford commented:
Looks like there is still 1 failing unit test but I don't think this is me as it looks like an IO exception accessing a temp file 🤷‍♂️

jzabroski commented:
I looked at the failing test: https://github.com/fluentmigrator/fluentmigrator/blob/9c87d698b161b3a1bb596a5624c02891a64459c5/test/FluentMigrator.Tests/IssueTests/GH0904/Fixture.cs

I think there are two possible reasons this could be failing:
1. IMigrationScopeStarter doesn't implement IDisposable, so there is no obvious point at which the Sqlite db file handle is released. - This might not even be the right interface "to blame", but you get the idea.
2. Microsoft.Data.Sqlite has a bug in the latest version we upgraded to?

mattbrailsford commented:
@jzabroski hmm, on investigation it looks like it's a known issue. By default SQLite will pool connections so calling `connection.Close()` won't release the DB file whilst pooling is enabled (the default). 

I've updated the connection string to disable pooling and that does indeed release the DB when the connection is closed.

There is also another suggested thing to do which is to force Garbage Collection but in my local test that didn't seem necessary once pooling was disabled.

mattbrailsford commented:
@jzabroski that unit test gave me an idea for getting some of the other unit tests working and it appeared to do the trick so now there are more unit tests passing for SQLite

jzabroski commented:
> n investigation it looks like it's a known issue. By default SQLite will pool connections so calling `connection.Close()` won't release the DB file in case it is needed again.
> 
> I've updated the connection string to disable pooling and that does indeed release the DB when the connection is closed.

That would probably be a good thing to add to the FluentMigrator/Documentation FAQ, under SQLite specific topics.

mattbrailsford commented:
@jzabroski feature wise? Yes, but it's the features that would require you to do the complex "Create new table, copy data, delete old table, rename new table" routine. I'm yet to figure out how we can safely do that. But anything that is natively supported by SQLite should all now be implemented.

jzabroski commented on src/FluentMigrator.Runner.SQLite/Generators/SQLite/SQLiteColumn.cs:
```diff
@@ -21,11 +22,43 @@ public override string Generate(IEnumerable<ColumnDefinition> columns, string ta
+        {
+            var fk2 = (ForeignKeyDefinition)foreignKey.Clone();
+
+            // SQLite FK's most be within the same schema as the FK itself
```
typo.  most -> must

mattbrailsford replied:
Updated comment 👍

jzabroski commented on src/FluentMigrator.Runner.SQLite/Generators/SQLite/SQLiteColumn.cs:
```diff
@@ -35,7 +68,7 @@ protected override string FormatIdentity(ColumnDefinition column)
                 if (!column.IsPrimaryKey && (!column.Type.HasValue || GetTypeMap(column.Type.Value, null, null) != "INTEGER"))
                 {
-                    throw new ArgumentException("SQLite only supports identity on a single integer, primary key coulmns");
+                    throw new ArgumentException("SQLite only supports identity on a single integer, primary key coulmn");
```
I wonder if we should also format this message to include the column name from `ColumnDefinition column` so that there is some context.  Not critical.

mattbrailsford replied:
Updated comment 👍

jzabroski commented on test/FluentMigrator.Tests/Integration/MigrationRunnerTests.cs:
```diff
@@ -629,6 +623,10 @@ public void CanMigrateASpecificVersionDown()
         {
             try
             {
+                // Can't currently test SQLite as we are using in-memory mode
```
@mattbrailsford  I believe this comment is no longer accurate?

mattbrailsford replied:
Removed comment

jzabroski commented on test/FluentMigrator.Tests/Integration/MigrationRunnerTests.cs:
```diff
@@ -1962,10 +1946,15 @@ internal class TestIndexNamingConventionWithSchema : Migration
-            var createSchemaExpr = Create.Schema("TestSchema");
+            // SQLite doesn't support creating schemas so for non SQLite DB's we'll create
+            // the schema, but for SQLite we'll attach a temp DB with the schema alias
+            var createSchemaExpr = IfDatabase(t => t != "SQLite").Create.Schema("TestSchema");
```
I was surprised to see we dont have a class with all the processor id's define. e.g.,
```c#
public static class ProcessorId
{
  SQLite = nameof(SQLite);
}
```
and then this code would read: 
```c#
var createSchemaExpr = IfDatabase(t => t != ProcessorId.SQLite).Create.Schema("TestSchema");
```
and SqliteProcessor would be updated so that line 50 is https://github.com/fluentmigrator/fluentmigrator/blob/9c87d698b161b3a1bb596a5624c02891a64459c5/src/FluentMigrator.Runner.SQLite/Processors/SQLite/SQLiteProcessor.cs#L50 :
```c#
public override string DatabaseType => ProcessorId.SQLite;
```

mattbrailsford replied:
Where would you see the `ProcessorId` class living as if it's in core it's not then extendable by custom provider implementations. Unless we are only really interested here in Core supported providers (which I think is reasonable)

mattbrailsford replied:
The `FluentMigrator` project made most sense to me so I've put that in place and updated the majority of places

jzabroski commented on test/FluentMigrator.Tests/Integration/MigrationRunnerTests.cs:
```diff
@@ -1980,14 +1969,23 @@ public override void Down()
-            Delete.Schema("TestSchema");
+            IfDatabase(t => t != "SQLite").Delete.Schema("TestSchema");
+
+            // Can't actually detatch SQLite DB here as migrations run in a transaction
```
~I don't think users should be attaching or detaching databases (schemas) in a migration.  I would just remove this comment and the commented out code unless I'm missing something.~ Nevermind.  I understand.  The Up migration is attaching as a side-effect.  It would be ideal of the attach happened at the connection-level.  But this test does raise an interesting scenario of being able to use IfDatabase to specialize behavior.

If pool is disabled, I assume that the database will auto-detach if running in TransactionPerMigration mode?

mattbrailsford replied:
I'm not entirely sure as the SQLite docs to clearly state the behaviour, but given attached databases are attached per connection, I would assume yes, that on closing the connection the attached database is auto-detached.

jzabroski reviewed (changes requested):
@mattbrailsford This mostly looks spot on to me.  There are some typos.  I also had a question about IntegrationTestBase and if there was a cleaner way to adjust the connection string through IntegrationTestOptions.SQLite.  It wasn't clear why this if statement was needed.


In a separate PR, do you have any interest in cleaning up the hard-coded Processor.DatabaseType values so that we have a constants file that defines all the processors?  

We can then look into @fairking  request to modularize FluentMigrator so we're not shipping every runner every time. It's probably the most requested feature right now.

mattbrailsford commented on src/FluentMigrator.DotNet.Cli/FluentMigrator.DotNet.Cli.csproj:
```diff
@@ -12,14 +12,15 @@
     <PackageReference Include="MySql.Data" Version="8.0.28" />
     <PackageReference Include="Npgsql" Version="3.2.7" />
     <PackageReference Include="Oracle.ManagedDataAccess.Core" Version="2.19.60" />
+    <PackageReference Include="SQLitePCLRaw.bundle_e_sqlite3" Version="2.0.7" />
```
Without it I get the error I mentioned previously in the notes section in the original post `Unable to load DLL 'e_sqlite3'` In my research, this was the advised solution and is what made it actually work.

mattbrailsford commented on src/FluentMigrator.Runner.SQLite/Generators/SQLite/SQLiteColumn.cs:
```diff
@@ -21,11 +22,43 @@ public override string Generate(IEnumerable<ColumnDefinition> columns, string ta
+            // As we generate FKs as part of the create table statement we need a way to prevent
+            // these FK's from creating FK constraints (which SQLite doesn't support) so we prefix
+            // the FK name and ignore anything in the "CreateConstraint" handler that has this name prefix
+            foreach (var fk in foreignKeyColumns) {
```
Not strictly a benefit, it's just that SQLite only supports FK's to be added in a CREATE TABLE statement which we are handling but FM really assumes that FK's will be added as constraints and so I need a way to signal to the constraint generator that the FK has been generated already and so don't error for those particular FK's. It's only if someone creates an explicit FK via the Constraints API that we want to throw a compatibility error.

If we don't ignore the FK's generated in the CREATE Table statement then FM attempts to also add them as contraints and so would throw an error.

mattbrailsford commented on test/FluentMigrator.Tests/FluentMigrator.Tests.csproj:
```diff
@@ -31,6 +31,7 @@
     <PackageReference Include="NUnit3TestAdapter" Version="3.17.0" />
     <PackageReference Include="Sap.Data.SQLAnywhere" Version="17.0.7.3399" />
     <PackageReference Include="Shouldly" Version="3.0.2" />
+    <PackageReference Include="SQLitePCLRaw.bundle_e_sqlite3" Version="2.0.7" />
```
See other response. Appears to be necessary to avoid an exception.

mattbrailsford commented on src/FluentMigrator.Runner.SQLite/Generators/SQLite/SQLiteQuoter.cs:
```diff
@@ -20,7 +20,21 @@ public override string FormatSystemMethods(SystemMethods value)
         public override string QuoteSchemaName(string schemaName)
         {
-            return string.Empty;
+            // SQLite doesn't support Schemas in the same sense as traditional SQL Server does,
```
Updated comment 👍

mattbrailsford commented on src/FluentMigrator.Runner.SQLite/Generators/SQLite/SQLiteColumn.cs:
```diff
@@ -45,17 +78,17 @@ protected override string FormatIdentity(ColumnDefinition column)
+        public override bool ShouldPrimaryKeysBeAddedSeparately(IEnumerable<ColumnDefinition> primaryKeyColumns)
         {
-            return column.IsPrimaryKey ? "PRIMARY KEY" : string.Empty;
+            //If there are no identity column then we can add as a separate constrint
```
Updated comment 👍

mattbrailsford commented on test/FluentMigrator.Tests/Integration/IntegrationTestBase.cs:
```diff
@@ -178,6 +192,11 @@ private bool IsAnyServerEnabled(Predicate<Type> isMatch)
+            var connectionString = serverOptions.ConnectionString;
+
+            if (processorType == typeof(SQLiteProcessor))
+                connectionString = $"Data Source={ _sqliteDbPath };Pooling=false;";
```
I did try setting the connection string within app settings to `Data Source=|Data Directory|\\TestDatabase.db;Pooling=False;` and in theory SQLite should run in a mode that creates the DB if it doesn't exist but for whatever reason I just kept getting an exception that it couldn't open the database.

The only way I could get it to work was to generate a temp DB file per test, putting it in the temp directory so it gets deleted on cleanup and then overriding the default connection string.

I can see if I can move this into a custom method for IntegrationTestOptions though.

mattbrailsford replied:
I've tried to improve this a bit by not making it SQLite specific. The issue I had was that the `Data Directory` doesn't get updated when in a test loop so rather than just handling SQLite specific code I've implemented something that will replace the Data Directory in the connection string to point to the tests temp directory as this ensures the DB is deleted every time.

This seams to have done the trick.

mattbrailsford commented:
Oops, just realized I did the ProcessorId thing as part of this PR, but hopefully that's ok. Sorry.

jzabroski commented:
> Oops, just realized I did the ProcessorId thing as part of this PR, but hopefully that's ok. Sorry, it's kinda bumped up the number of files changed.

that's ok

jzabroski commented:
Thanks for your hard work, @mattbrailsford

mattbrailsford commented:
No problem. Hope you / the community are happy with it 👍🏻
